dataset.py

```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  # Sort by polar angle. This arranges the coordinates so that they
+  # delimit an outward shape.
+  sorted_coordinates = list(coordinates)
+  sorted_coordinates.sort(
+    key=lambda c: math.atan2(
+      c[1] - centroid[1],
+      c[0] - centroid[0])
+  )
+  return Polygon(sorted_coordinates)
+
+def _valid_polygons(
+  train_polygon: Polygon,
+  validation_polygon: Polygon,
+  test_polygon: Polygon) -> bool:
+  '''
+  Returns true if the train, validation, and test polygons
+  have a larger area than or equal to 0.1 and they don't overlap
+  over each other. Returns false otherwise. 
+  '''
+  small_polygon = False
+  for polygon in [train_polygon, validation_polygon, test_polygon]:
+    if polygon.area < 0.1:
+      return False
+
+  if (train_polygon.intersects(validation_polygon) or
+     train_polygon.intersects(test_polygon) or
+     validation_polygon.intersects(test_polygon)):
+    return False
+
+  return True
+
+def _shuffled_unique_coordinates(
+  sample_data: pd.DataFrame,
+  strategy: FurthestPointsPartitionStrategy) -> list[list[float]]:
+  '''
+  Returns shuffled unique coordinates from sample_data that match
+  the expected train, validation and test fractions.
+  '''
+  unique_coordinates_df = sample_data.groupby(
+    by=['long', 'lat']
+  ).first().reset_index()
+  shuffled_coordinates_df = unique_coordinates_df.sample(frac=1.0,
+    random_state=strategy.random_seed)
+  coordinates = list(
+    shuffled_coordinates_df[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+  assert(len(coordinates) >= strategy.top_n)
+  min_size = (int(len(coordinates) * strategy.train_fraction) +
+    int(len(coordinates) * strategy.validation_fraction) +
+    int(len(coordinates) * strategy.test_fraction))
+  assert(int(len(coordinates) * strategy.train_fraction) > 0)
+  assert(int(len(coordinates) * strategy.validation_fraction) > 0)
+  assert(int(len(coordinates) * strategy.test_fraction) > 0)
+  assert(
+    min_size <= len(coordinates)
+  ), (f"You need {min_size} aka {int(len(coordinates) * strategy.train_fraction)} train + "
+    f"{int(len(coordinates) * strategy.validation_fraction)} validation + "
+    f"{int(len(coordinates) * strategy.test_fraction)} test samples but you have a sample of {len(coordinates)}")
+  
+  return coordinates
+
+def _maybe_partition_furthest_points(
+  sample_data: pd.DataFrame,
+  furthest_coordinates: list[list[float]],
+  strategy: FurthestPointsPartitionStrategy) -> PartitionedDataset:
+  '''
+  Returns a PartitionedDataset that has a valid train/validation/test split
+  that has:
+  - Non-overlapping splits
+  - Splits which polygon areas have less than 0.1 area
+  It attempts to generate these partitions randomly for at most strategy.max_attempts.
+  If unsuccesful, returns None.
+  '''
+  partitioned_dataset = None
+  attempts = 0
+  while (attempts <= strategy.max_attempts and
+         partitioned_dataset is None):
```

Comment:
nit: A more Pythonic way to do this might be `for attempt in range(strategy.max_attempts):.... break`. When I first came to Python from C++, I would usually do this instead so in a way I'm just passing on feedback from others :-).

Either way, I think people will understand what you mean. But in the Python community, we tend to be faster at reading the `for` loop variant of this.

Reply:
Sounds good. Updated.